pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,235 @@
+package alertmanager
+
+import (
+	"context"
+	"hash/fnv"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/client"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Distributor forwards requests to individual alertmanagers.
+type Distributor struct {
+	services.Service
+
+	cfg              *MultitenantAlertmanagerConfig
+	requestsInFlight sync.WaitGroup
+
+	alertmanagerRing        ring.ReadRing
+	alertmanagerClientsPool ClientsPool
+
+	logger log.Logger
+}
+
+// NewDistributor constructs a new Distributor
+func NewDistributor(cfg *MultitenantAlertmanagerConfig, alertmanagersRing *ring.Ring, alertmanagerClientsPool ClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
```

Comment:
Let's extract options used by Distributor to its own Config structure. No need to pass entire unrelated config structure to Distributor.

Reply:
Turns out we only need `cfg.AlertmanagerClient`, so instead of creating a new distributor config, I will pass this one instead